aip/0162.md

```diff
@@ -0,0 +1,229 @@
+
+### Listing revisions
+
+APIs implementing resource revisions **should** provide a custom method for
+listing the revision history for a resource, with a structure similar to
+standard `List` methods ([AIP-132][]):
+
+```proto
+rpc ListBookRevisions(ListBookRevisionsRequest)
+    returns (ListBookRevisionsResponse) {
+  option (google.api.http) = {
+    get: "/v1/{name=publishers/*/books/*}:listRevisions"
+  }
+}
+```
+
+```proto
+message ListBookRevisionsRequest {
+  // The name of the book to list revisions for.
+  string name = 1;
+
+  // The maximum number of revisions to return per page.
+  int32 page_size = 2;
+
+  // The page token, received from a previous ListBookRevisions call.
+  // Provide this to retrieve the subsequent page.
+  string page_token = 3;
+}
+
+message ListBookRevisionsResponse {
+  // The revisions of the book.
+  repeated BookRevision book_revisions = 1;
+
+  // A token that can be sent as `page_token` to retrieve the next page.
+  // If this field is omitted, there are no subsequent pages.
+  string next_page_token = 2;
+}
+```
+
+While revision listing methods are mostly similar to standard `List` methods
+([AIP-132][]), the following important differences apply:
+
+- The first field in the request message **must** be called `name` rather than
+  `parent` (this is listing revisions for a specific book, not a collection of
+  books).
+- The URI **must** end with `:listRevisions`.
+- Revisions **must** be ordered in reverse chronological order.
+
+The revision object **should** only be used by the revision list method:
+
+```proto
+message BookRevision {
+  // A representation of the book at this revision.
+  // The `name` field always includes an explicit revision ID.
+  Book book = 1;
+
+  // The date that this revision was created.
+  google.protobuf.Timestamp create_time = 2;
+}
+```
+
+- Revision objects **should** include a representation of the resource as of
+  that revision.
+  - The `string name` field in the resource **must** include an explicit
+    revision ID.
+  - If providing the full resource is expensive or infeasible, the revision
+    object **may** include a `string name` field instead, which **must**
+    include the resource name and an explicit revision ID.
+- Revision objects **must** include the timestamp when the revision was
+  created, which **should** be called `create_time`.
+- Revision objects **may** include other information, such as a description or
+  other metadata, if desired.
+
+### Snapshotting revisions
+
+Depending on the resource, different APIs may have different strategies for
+when to snapshot a new revision, such as:
+
+- Snapshot a new revision any time that there is a change
+- Snapshot a new revision when something important happens
+- Snapshot a new revision when the user specifically asks
+
+APIs **may** use any of these strategies. APIs that want to snapshot a revision
+on user request **should** handle this with a `Snapshot` custom method:
```

Comment:
Might need to rename this because we have been using the term `Snapshot` to have another meaning.

Reply:
Changed to `Commit`.